sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -498,7 +498,8 @@ private BytesReference prepareInsert(DocTableInfo tableInfo,
                 indexShard.shardId(), item.opType(), item.id(), version, item.source().utf8ToString());
         }
         return indexShard.prepareIndexOnPrimary(
-            sourceToParse, version, item.versionType(), -1, request.isRetry());
+            // always use internal version type for primary operations!
```

Comment:
I'd remove the comment or change it to something along the lines of `must not use item.versionType(); it might have been mutated to VersionType.EXTERNAL if we're here due to a retry`

The current comment is in an exaggerated sense akin to "usually it doesn't matter what kind of arguments we pass along, but here it really matters and I'm not going to tell you why"

Reply:
Good point, "why" type comments are much more helpful.